sched/scheduler/stateful_scheduler.go

```diff
@@ -355,39 +363,53 @@ func (s *statefulScheduler) scheduleTasks() {
 			taskId: taskId,
 			task:   taskDef,
 			nodeId: nodeId,
+
+			abortCh: make(chan interface{}, 1),
 		}
 
+		// mark the task as started in the jobState and record its taskRunner
+		jobState.taskStarted(taskId, tRunner)
+
 		s.asyncRunner.RunAsync(
-			run.run,
+			tRunner.run,
 			func(err error) {
+				log.Infof("In task callback for job: Error:%+v", err)
 				flaky := false
 				if err != nil {
 					// Get the type of error. Currently we only care to distinguish runner (ex: thrift) errors to mark flaky nodes.
 					taskErr := err.(*taskError)
+					st := err.(*taskError).st
+					aborted := (st.State == runner.ABORTED)
 					flaky = (taskErr.runnerErr != nil)
 
 					msg := "Error running job (will be retried):"
-					if taskErr.resultErr != nil && taskErr.st.State == runner.COMPLETE {
-						msg = "Error running job (quitting, tasks that run to completion are not retried):"
-						err = nil
-					} else if preventRetries {
-						msg = fmt.Sprintf("Error running job (quitting, hit max retries of %d):", s.maxRetriesPerTask)
-						err = nil
-					} else {
-						jobState.errorRunningTask(taskId, err)
+					if aborted {
+						msg = "Task aborted."
+					}
+					if !aborted {
+						if preventRetries {
+							msg = fmt.Sprintf("Error running job (quitting, hit max retries of %d):", s.maxRetriesPerTask)
+							err = nil
+						} else {
+							jobState.errorRunningTask(taskId, err)
+						}
 					}
 					log.Info(msg, jobId, ", task:", taskId, " err:", taskErr, " cmd:", taskDef.Argv)
 
 					// If the task completed succesfully but sagalog failed, start a goroutine to retry until it succeeds.
 					if taskErr.sagaErr != nil && taskErr.runnerErr == nil && taskErr.resultErr == nil {
 						log.Info(msg, jobId, ", task:", taskId, " -> starting goroutine to handle failed saga.EndTask. ")
+						//TODO -this may results in closed channel panic due to sending endSaga to sagalog (below) before endTask
 						go func() {
-							for err := errors.New(""); err != nil; err = run.logTaskStatus(&taskErr.st, saga.EndTask) {
+							for err := errors.New(""); err != nil; err = tRunner.logTaskStatus(&taskErr.st, saga.EndTask) {
 								time.Sleep(time.Second)
 							}
 							log.Info(msg, jobId, ", task:", taskId, " -> finished goroutine to handle failed saga.EndTask. ")
 						}()
 					}
+					if aborted {
```

Comment:
Why is this a separate block from the "if aborted" above? And can you just say "err == nil || aborted" below?